README.md
# Echarts Scraper Puppeteer

[![Build Status](https://travis-ci.org/chfw/echarts-scrappeteer.svg?branch=master)](https://travis-ci.org/chfw/echarts-scrappeteer) [![codecov](https://codecov.io/gh/chfw/echarts-scrappeteer/branch/master/graph/badge.svg)](https://codecov.io/gh/chfw/echarts-scrappeteer) [![npm version](https://badge.fury.io/js/echarts-scrappeteer.svg)](https://badge.fury.io/js/echarts-scrappeteer)

For svg output, it is required that **echarts 3.8+** is used and the echarts instance
is [renderred as svg](https://ecomfe.github.io/echarts-doc/public/en/tutorial.html#Render%20by%20Canvas%20or%20SVG).

## Introduction

It scrapes all echarts found in a web page as images. And it scrapes 3D charts too. Of course, it supports
[pyecharts](https://github.com/chenjiandongx/pyecharts) as well. Please find the example in later section.

![demo](https://github.com/chfw/echarts-scrappeteer/raw/master/demo.gif)

## Installation

```shell
$ npm i -g echarts-scrappeteer
```

### requirements

Node 7.6.0 or later

## Usage

```shell

Usage: scrappeteer [options] <url/file>


Options:

  -f, --format <png,jpeg,gif,svg>         image format
  -o, --output <outputname>               output file name
  -w, --wait <delay in milli-seconds>     wait a while before scrapping
  -v, --viewPort <width,height>           force puppeteer to set viewport. for echarts gallery site only
  -r, --clipRectangle <x,y,width,height>  record rectangle when making gif animation
  -c, --frameCounts <number>              of frames. gif only
  -i, --frameInterval <number>            frame intervals. gif only
  -s, --skipFrames <number>               skip initial frames. gif only
  -g, --gap time<number>                  between each gif snapshot. gif only
  -h, --help                              output usage information
```

If the page load speed is slow or if the resulting image is partial, `-w` parameter is
required to delay the scrape action.

Please note that if more than one echarts should be found in a page, `-f` option applies to all.

## Example

### Scrape echarts from a URL

Here is the command for your discretion:

```
$ scrappeteer https://pyecharts.github.io/echarts-china-cities-js/preview.html
```

### Scrape echarts from a local file


```shell
$ scrappeteer render.html
```

Where does the `render.html` come from? It is genereted by pyecharts. Please visit [pyecharts-demo.py](https://github.com/chfw/echarts-scrappeteer/blob/master/pyecharts-demo/pyecharts-demo.py). 


![scrape pyecharts](https://github.com/chfw/echarts-scrappeteer/raw/master/scraped-gallery/output.4.png)


## Gallery

### Gif format :fire:

```shell
scrappeteer http://gallery.echartsjs.com/editor.html?c=xrkJtnKJq- -w 2000 -f gif -r 525,50,770,750 -o gf3 -i 300 -c 10 -v 1300,800
```

[![high speed route](https://github.com/chfw/echarts-scrappeteer/raw/master/scraped-gallery/shenzhen.gif)](http://gallery.echartsjs.com/editor.html?c=xrkJtnKJq-)

### Static images

[![step count](https://github.com/chfw/echarts-scrappeteer/raw/master/scraped-gallery/step-count.png)](http://gallery.echartsjs.com/editor.html?c=calendar-effectScatter)
[![pig plan](https://github.com/chfw/echarts-scrappeteer/raw/master/scraped-gallery/pig-plan.png)](http://gallery.echartsjs.com/editor.html?c=xByUX8HuDZ)
[![surface wave](https://github.com/chfw/echarts-scrappeteer/raw/master/scraped-gallery/surface-wave.png)](http://gallery.echartsjs.com/editor.html?c=xHkcYXm9pe)

## LICENSE

ISC


CHANGELOG.md
# Change log

## 1.2.0 - 9.1.2018

### Added

1. [#1](https://github.com/chfw/echarts-scrappeteer/issues/1) support svg output in echarts 3.8.x.

## 1.1.1 - 21.9.2017

### Fixed

1. Fix image quality issue under mac
1. Fix default wait issue in 1.1.0. It is now 30 seconds.

## 1.1.0 - 19.9.2017

### Added

Support scraping gif format for echarts on http://gallery.echartsjs.com/

## 1.0.1 - 9.9.2017

### Added

Support scraping echarts on http://gallery.echartsjs.com/

### Updated

Use puppeteer 0.10.2

## 1.0.0

Initial support


test/main.spec.js
const rewire = require('rewire');
const assert = require('assert');
const path = require('path');
const scrappeteer = require(path.join(process.cwd(), 'src', 'scrappeteer.js'));
const sinon = require('sinon');

var main = rewire(path.join(process.cwd(), 'src', 'main.js'));
var main_func = main.__get__('main');
var snapshot = sinon.stub(scrappeteer, 'snapshot');
snapshot.returns(0);

/* test mockup */
var sandbox   = sinon.sandbox.create({ useFakeTimers : true });
var exitStub  = sandbox.stub(process, 'exit');


describe('main', function(){

  it('should handle a url', function(){
    main_func('file', {format: 'png',
                       output: 'output'});
    assert(snapshot.called);
  });

  it('should handle undefined parameters', function(){
    main_func('file', {});
    assert(snapshot.called);
  });

  it('should accept clip rectangle', function(){
    main_func('file', {format: 'jpeg', output: 'output', clipRectangle: [1, 2, 3, 4]});
  });

  it('should reject unsupported format', function(){
    process.once('SIGTERM', () => {
      closeStub = sandbox.stub(server, 'close');
    });
    main_func('file', { format: 'tif', output: 'output' });
  });

  it('should reject invalid view port', function(){
    process.once('SIGTERM', () => {
      closeStub = sandbox.stub(server, 'close');
    });
    main_func('file', {format: 'jpeg', output: 'output', viewPort: [1, 2, 3, 4]});
  });

});


describe('main utils', function(){
  it('should convert number string into number', function(){
    const intValue = main.__get__('intValue');
    var result = intValue('1000');
    assert.equal(result, 1000);
  });

  it('should convert number string array into number array', function(){
    const intArray = main.__get__('intArray');
    var result = intArray('1,2,3');
    assert.equal(result[0], 1);
    assert.equal(result[1], 2);
    assert.equal(result[2], 3);
  });

  it('should get default view port', function(){
    const viewPort = main.__get__('viewPort');
    var ret = viewPort('1,2');
    assert.equal(ret[2], 1);
  });
});


test/scrappeteer.spec.js
const rewire = require('rewire');
const fs = require('fs');
const path = require('path');
var assert = require('assert');
const sinon = require('sinon');


var scrappeteer = rewire(path.join(process.cwd(), 'src', 'scrappeteer.js'));

const p = require('puppeteer');
var browser = {
  newPage: () => {
    var page = {
      goto: sinon.spy(),
      waitForNavigation: sinon.spy(),
      setViewport: sinon.spy(),
      mainFrame: function(){
	var mainFrame = {
	  childFrames: function(){
	    return ['fake child frame'];
	  }
	}
	return mainFrame;
      }
    };

    return page;
  },
  close: sinon.spy()
}
var launch = sinon.stub(p, 'launch');
launch.returns(browser);


describe('Scrappeteer', function(){

  it('should save data url as image', function(){
    var saveDataUrl = scrappeteer.__get__('saveDataUrl');
    var testFixture = path.join(process.cwd(), 'test', 'google.data.url.txt');
    fs.readFile(testFixture, "utf8", function read(err, data){
      if (err) {
	throw err;
      }
      saveDataUrl(data.trim(), 1, 'output');
      var status = fs.existsSync('output.1.png');
      assert.equal(status, true);
      fs.unlink('output.1.png', (err) => {});
    });
  });

  it('should count the number of charts', function(){
    var countCharts = scrappeteer.__get__('countCharts');
    var page = {
      evaluate: () => {}
    };
    var stub = sinon.stub(page, 'evaluate');
    stub.onCall(0).resolves(1);

    countCharts(page).then( (count)=>{
      assert.equal(count, 1);
    }).catch((e)=>{});

  });

  it('should get a chart', function(){
    var getAChart = scrappeteer.__get__('getAChart');
    var page = {
      evaluate: () => {}
    };
    var stub = sinon.stub(page, 'evaluate');
    stub.onCall(0).resolves(1);

    getAChart(page, 'png', 1).then( (count)=>{
      assert.equal(count, 1);
    });
  });

  it('should get a svg chart', function(){
    var getAChart = scrappeteer.__get__('getAChart');
    var page = {
      evaluate: () => {}
    };
    var stub = sinon.stub(page, 'evaluate');
    stub.onCall(0).resolves(1);

    getAChart(page, 'svg', 1).then( (count)=>{
      assert.equal(count, 1);
    });
  });

  it('should record gif animation', function(){
    var recordGif = scrappeteer.__get__('recordGif');
    var page = {
      screenshot: function(){
        return fs.readFileSync(path.join(process.cwd(), 'test', 'test.png'));
      },
      waitForNavigation: function(x){}
    }
    var options = {
      clipRect: {
        width: 210,
        height: 211
      },
      outputName: "giftest",
      frameCounts: 2,
      snapshotInterval: 1
    }
    recordGif(page, options);
  });

  it('should scrape many echarts', function(){
    var scrapeEcharts = scrappeteer.__get__('scrapeEcharts');
    /* Note: the following three statement is permanent */
    scrappeteer.__set__('countCharts', sinon.stub().returns(10));
    scrappeteer.__set__('getAChart', sinon.stub().returns(0));
    scrappeteer.__set__('saveDataUrl', sinon.stub().returns(0));
    scrapeEcharts('fake page', {imageFormat:'png', outputName:'output'});
  });


  describe('should take snapshots', function(){

    it('from normal file/link', function(){
      scrappeteer.snapshot('url', {imageFormat:'png', outputName:'output'});
    });

    it('on echarts gallery', function(){
      scrappeteer.snapshot('http://gallery.echartsjs.com/echart', 'png', 'output');
    });

    it(' of gif animation', function(){
      scrappeteer.__set__('recordGif', sinon.stub().returns(0));
      scrappeteer.snapshot('url', {imageFormat:'gif', outputName:'output'});
    });

  });

});


index.js
#!/usr/bin/env node

const p = require('puppeteer');
const path = require('path');
const fs = require('fs');

if(process.argv.length < 4){
	console.log("Usage: " + __filename + " pyecharts_outpout.html [png|jpeg]");
	process.exit(-1);
}

var html_file = process.argv[2];
const image_format = process.argv[3];

if(html_file.indexOf("http") == -1 and html_file.indexOf("file://") == -1){
	html_file = 'file://' + path.join(__dirname, html_file);
}

(async () => {

	const browser = await p.launch({headless: false});
	const page = await browser.newPage();
	await page.goto(html_file, {waitUntil: 'networkidle'});

	const number = await page.evaluate((args) => {
		var ele =  document.querySelectorAll('div[_echarts_instance_]');
		return ele.length;
	}, {image_format: image_format});

	console.log("Found " + number + " echarts");
	for(i=0;i<number;i++){
		const result = await page.evaluate((args) => {
			var ele =  document.querySelectorAll('div[_echarts_instance_]');
			var mychart = echarts.getInstanceByDom(ele[args.index]);
			return mychart.getDataURL({type:args.image_format,
									   excludeComponents: ['toolbox']});
		}, {image_format: image_format, index: i});

		var regex = /^data:.+\/(.+);base64,(.*)$/;

		var matches = result.match(regex);
		var ext = matches[1];
		var data = matches[2];
		var buffer = new Buffer(data, 'base64');
		fs.writeFileSync('output.' + i + '.' + ext, buffer);
	}
	
	browser.close();

})();


src/scrappeteer.js
const fs = require('fs');
const os = require('os');
const p = require('puppeteer');
const path = require('path');
const ProgressBar = require('progress');
const chalk = require('chalk');

var GIFEncoder = require('gifencoder');
var PNG = require('png-js');


const ECHARTS_GALLERY = 'http://gallery.echartsjs.com/';


var takeSnapshots = (async (urlOrFile, options) => {
  if(urlOrFile.indexOf("http") == -1 && urlOrFile.indexOf("file://") == -1){
    urlOrFile = 'file://' + path.join(process.cwd(), urlOrFile);
  }

  const browser = await p.launch({headless: false, timeout: options.wait});
  const page = await browser.newPage();
  if(urlOrFile.indexOf(ECHARTS_GALLERY) != -1){
    await page.setViewport(options.viewPort);
  }
  page.goto(urlOrFile,{waitUtil: 'networkidle'});

  try{
    await page.waitForNavigation({timeout: options.wait});
  }catch(e){
  }

  if(options.imageFormat === 'gif'){
    await recordGif(page, options);
  }else{
    if(urlOrFile.indexOf(ECHARTS_GALLERY) != -1){
      const mainFrame = page.mainFrame();
      const childFrames = mainFrame.childFrames();
      await scrapeEcharts(childFrames[0], options.imageFormat, options.outputNAme)
    } else {
      await scrapeEcharts(page, options.imageFormat, options.outputName);
    }
  }
  browser.close();

});


async function recordGif(page, options){
  var encoder = new GIFEncoder(options.clipRect.width, options.clipRect.height);
  encoder.createWriteStream()
    .pipe(fs.createWriteStream(options.outputName + '.gif'));

  encoder.start();
  encoder.setRepeat(0);   // 0 for repeat, -1 for no-repeat
  encoder.setDelay(options.frameInterval);  // frame delay in ms
  encoder.setQuality(10); // image quality. 10 is default.

  for(i=0;i<options.frameCounts;i++){
    var pngBuffer = await page.screenshot({clip: options.clipRect});
    if(i > options.skipFrames){
      var png = new PNG(pngBuffer);
      png.decode(function(pixels){
        encoder.addFrame(pixels);
      });
    }
    try{
      await page.waitForNavigation({timeout: options.snapshotInterval});
    }catch(e){
    }

  }
  encoder.finish();
}


async function scrapeEcharts(document, imageFormat, outputName){
  const numberOfCharts = await countCharts(document);
  console.log(chalk.cyan("Found ") + chalk.bold.green(numberOfCharts) +
              chalk.cyan(" echarts."));

  var barOpts = {
    width: 20,
    total: numberOfCharts,
    clear: true
  };
  var bar = new ProgressBar(chalk.cyan('Scrappinging [:bar] :percent :etas'),
                            barOpts);
  for(i=0; i<numberOfCharts; i++){
    const dataurl = await getAChart(document, imageFormat, i);
    if(imageFormat === 'svg'){
      saveSvgFile(dataurl, i, outputName);
    }else{
      saveDataUrl(dataurl, i, outputName);
    }
    bar.tick(1);
  }

};


var __count_charts__ = () => {
  var echarts =  document.querySelectorAll('div[_echarts_instance_]');
  return echarts.length;
}

var __get_chart__ = (args) => {
  var ele =  document.querySelectorAll('div[_echarts_instance_]');
  var mychart = echarts.getInstanceByDom(ele[args.index]);
  return mychart.getDataURL({
    type:args.image_format,
    pixelRatio: args.devicePixelRatio,
    excludeComponents: ['toolbox']});
}

var __get_svg_chart__ = (args) => {
  var ele =  document.querySelectorAll('div[_echarts_instance_] svg');
  return ele[args.index].outerHTML;
}

async function countCharts(page){
  return await page.evaluate(__count_charts__);
}


async function getAChart(page, imageFormat, index){
  var args = {
    image_format: imageFormat,
    index: index,
    devicePixelRatio: 1
  }
  var is_mac = os.platform() === 'darwin';
  if(is_mac){
    args.devicePixelRatio = 2;
  }
  
  if (imageFormat === "svg"){
    return page.evaluate(__get_svg_chart__, args);
  }else{
    return page.evaluate(__get_chart__, args);
  }
}


function saveDataUrl(dataurl, index, outputName){
  var regex = /^data:.+\/(.+);base64,(.*)$/;

  var matches = dataurl.match(regex);
  var ext = matches[1];
  var data = matches[2];
  var buffer = new Buffer(data, 'base64');
  fs.writeFileSync(outputName + '.' + index + '.' + ext, buffer);
}


function saveSvgFile(dataurl, index, outputName){
  fs.writeFileSync(outputName + '.' + index + '.svg', dataurl);
}

module.exports = {
  snapshot: takeSnapshots
};


pyecharts-demo/pyecharts-demo.py
# coding=utf-8

# Before you try the example code, please do
#   pip install pyecharts

from __future__ import unicode_literals

import os
import random
from pyecharts import (
    Scatter3D, Line, Pie,
    Kline, Radar, WordCloud, Liquid)
from pyecharts import Page


RANGE_COLOR = [
    '#313695', '#4575b4', '#74add1', '#abd9e9', '#e0f3f8',
    '#ffffbf', '#fee090', '#fdae61', '#f46d43', '#d73027', '#a50026'
]


page = Page(page_title="Echarts Scrappeteer Example")
# line
attr = ['周一', '周二', '周三', '周四', '周五', '周六', '周日']
line = Line("折线图示例")
line.add("最高气温", attr, [11, 11, 15, 13, 12, 13, 10],
         mark_point=["max", "min"], mark_line=["average"])
line.add("最低气温", attr, [1, -2, 2, 5, 3, 2, 0],
         mark_point=["max", "min"], mark_line=["average"])
page.add(line)

# pie
attr = ["衬衫", "羊毛衫", "雪纺衫", "裤子", "高跟鞋", "袜子"]
v1 = [11, 12, 13, 10, 10, 10]
pie = Pie("饼图-圆环图示例", title_pos='center')
pie.add("", attr, v1, radius=[40, 75], label_text_color=None,
        is_label_show=True, legend_orient='vertical', legend_pos='left')
page.add(pie)

# kline
v1 = [[2320.26, 2320.26, 2287.3, 2362.94],
      [2300, 2291.3, 2288.26, 2308.38],
      [2295.35, 2346.5, 2295.35, 2345.92],
      [2347.22, 2358.98, 2337.35, 2363.8],
      [2360.75, 2382.48, 2347.89, 2383.76],
      [2383.43, 2385.42, 2371.23, 2391.82],
      [2377.41, 2419.02, 2369.57, 2421.15],
      [2425.92, 2428.15, 2417.58, 2440.38],
      [2411, 2433.13, 2403.3, 2437.42],
      [2432.68, 2334.48, 2427.7, 2441.73],
      [2430.69, 2418.53, 2394.22, 2433.89],
      [2416.62, 2432.4, 2414.4, 2443.03],
      [2441.91, 2421.56, 2418.43, 2444.8],
      [2420.26, 2382.91, 2373.53, 2427.07],
      [2383.49, 2397.18, 2370.61, 2397.94],
      [2378.82, 2325.95, 2309.17, 2378.82],
      [2322.94, 2314.16, 2308.76, 2330.88],
      [2320.62, 2325.82, 2315.01, 2338.78],
      [2313.74, 2293.34, 2289.89, 2340.71],
      [2297.77, 2313.22, 2292.03, 2324.63],
      [2322.32, 2365.59, 2308.92, 2366.16],
      [2364.54, 2359.51, 2330.86, 2369.65],
      [2332.08, 2273.4, 2259.25, 2333.54],
      [2274.81, 2326.31, 2270.1, 2328.14],
      [2333.61, 2347.18, 2321.6, 2351.44],
      [2340.44, 2324.29, 2304.27, 2352.02],
      [2326.42, 2318.61, 2314.59, 2333.67],
      [2314.68, 2310.59, 2296.58, 2320.96],
      [2309.16, 2286.6, 2264.83, 2333.29],
      [2282.17, 2263.97, 2253.25, 2286.33],
      [2255.77, 2270.28, 2253.31, 2276.22]]
kline = Kline("K 线图示例")
kline.add("日K", ["2017/7/{}".format(i + 1) for i in range(31)],
          v1, is_datazoom_show=True)
page.add(kline)

# radar
schema = [
    ("销售", 6500), ("管理", 16000), ("信息技术", 30000),
    ("客服", 38000), ("研发", 52000), ("市场", 25000)
]
v1 = [[4300, 10000, 28000, 35000, 50000, 19000]]
v2 = [[5000, 14000, 28000, 31000, 42000, 21000]]
radar = Radar("雷达图示例")
radar.config(schema)
radar.add("预算分配", v1, is_splitline=True, is_axisline_show=True)
radar.add("实际开销", v2, label_color=["#4e79a7"], is_area_show=False,
          legend_selectedmode='single')
page.add(radar)

# scatter3d
data = [
    [random.randint(0, 100),
     random.randint(0, 100),
     random.randint(0, 100)] for _ in range(80)
]
scatter3D = Scatter3D("3D 散点图示例", width=1200, height=600)
scatter3D.add("", data, is_visualmap=True, visual_range_color=RANGE_COLOR)
page.add(scatter3D)

# wordcloud
name = [
    'Sam S Club', 'Macys', 'Amy Schumer', 'Jurassic World',
    'Charter Communications', 'Chick Fil A', 'Planet Fitness',
    'Pitch Perfect', 'Express', 'Home', 'Johnny Depp', 'Lena Dunham',
    'Lewis Hamilton', 'KXAN', 'Mary Ellen Mark', 'Farrah Abraham',
    'Rita Ora', 'Serena Williams', 'NCAA baseball tournament', 'Point Break'
]
value = [
    10000, 6181, 4386, 4055, 2467, 2244, 1898, 1484, 1112,
    965, 847, 582, 555, 550, 462, 366, 360, 282, 273, 265
]
wordcloud = WordCloud(width=1300, height=620)
wordcloud.add("", name, value, word_size_range=[30, 100], rotate_step=66)
page.add(wordcloud)

# liquid
liquid = Liquid("水球图示例")
liquid.add("Liquid", [0.6])
page.add(liquid)
page.render()

os.system('scrappeteer render.html')


src/main.js
#!/usr/bin/env node

var program = require('commander');
const scrappeteer = require('./scrappeteer');
const chalk = require('chalk');


const DEFAULTS = {
  outputName: "output",
  wait: 30000,
  frameCounts: 1,
  frameInterval: 500,
  viewPort: [1300, 800, 1],
  imageFormat: 'png',
}


program
  .arguments('<url/file>')
  .option('-f, --format <png,jpeg,gif,svg>', 'image format')
  .option('-o, --output <outputname>', 'output file name')
  .option('-w, --wait <delay in milli-seconds>', 'wait a while before scrapping', intValue)
  .option('-v, --viewPort <width,height>', 'force puppeteer to set viewport. for echarts gallery site only', viewPort)
  .option('-r, --clipRectangle <x,y,width,height>', 'record rectangle when making gif animation', intArray)
  .option('-c, --frameCounts <number>', 'of frames. gif only', intValue)
  .option('-i, --frameInterval <number>', 'frame intervals. gif only', intValue)
  .option('-s, --skipFrames <number>', 'skip initial frames. gif only', intValue)
  .option('-g, --gapTime<number>', 'between each gif snapshot. gif only', intValue)
  .action(function(url_or_file){
    main(url_or_file, program);
  })
  .parse(process.argv);


function main(url_or_file, program){
  var options = {}
  if(program.frameCounts){
    options.frameCounts = program.frameCounts;
  }
  if(program.gapTime){
    options.snapshotInterval = program.gapTime;
  }
  if(program.frameInterval){
    options.frameInterval = program.frameInterval;
  }
  if(program.skil){
    options.skipFrames = program.skip;
  }
  if(program.wait){
    options.wait = program.wait;
  }
  if(program.format){
    options.imageFormat = program.format;
  }
  if(program.output){
    options.outputName = program.output;
  }
  if(program.clipRectangle){
    options.clipRect = {
      x: program.clipRectangle[0],
      y: program.clipRectangle[1],
      width: program.clipRectangle[2],
      height: program.clipRectangle[3]
    }
  }
  if(program.viewPort){
    if( program.viewPort.length != 3){
      console.error(chalk.cyan("Wrong view port parameter :") +
		    chalk.bold.red(program.viewPort))
      process.exit(1);
    }
    options.viewPort = {
      width: program.viewPort[0],
      height: program.viewPort[1],
      deviceScaleFactor: 1
    }
  }
  var final_options = Object.assign({}, DEFAULTS, options);
  if (final_options.imageFormat != 'png' && final_options.imageFormat != 'jpeg' && final_options.imageFormat != 'gif' && final_options.imageFormat != 'svg'){
    console.error(chalk.cyan("Unsupported file format : ") +
                  chalk.bold.red(final_options.imageFormat));
    process.exit(1);
  }
  scrappeteer.snapshot(url_or_file, final_options);
}


function intArray(val) {
  return val.split(',').map(function(x){ return parseInt(x, 10) });
}


function intValue(val) {
  return parseInt(val, 10);
}


function viewPort(val) {
  var ret = intArray(val);
  ret.push(1);
  return ret;
}
